File: DSA/src/Questions/BinarySearch/FindMinInRotatedSortedArray.java
package Questions.BinarySearch;

/*
 https://leetcode.com/problems/find-minimum-in-rotated-sorted-array-ii/description/
 Suppose an array of length n sorted in ascending order is rotated between 1 and n times. For example, the array nums = [0,1,4,4,5,6,7] might become:

[4,5,6,7,0,1,4] if it was rotated 4 times.
[0,1,4,4,5,6,7] if it was rotated 7 times.
Notice that rotating an array [a[0], a[1], a[2], ..., a[n-1]] 1 time results in the array [a[n-1], a[0], a[1], a[2], ..., a[n-2]].

Given the sorted rotated array nums that may contain duplicates, return the minimum element of this array.

You must decrease the overall operation steps as much as possible.

 

Example 1:

Input: nums = [1,3,5]
Output: 1
Example 2:

Input: nums = [2,2,2,0,1]
Output: 0
 

Constraints:

n == nums.length
1 <= n <= 5000
-5000 <= nums[i] <= 5000
nums is sorted and rotated between 1 and n times.
 Time Complexity: O(logN)
 Space Complexity: O(1)
 */
public class FindMinInRotatedSortedArray {
    public int findMin(int[] nums) {
        int left = 0;
        int right = nums.length - 1;

        while (left < right) {
            int mid = (right + left) >> 1;

            if (nums[mid] < nums[right]) {
                right = mid;
            } else if (nums[mid] > nums[right]) {
                left = mid + 1;
            } else {
                right--; // nums[mid] == nums[right], reduce the search space
            }
        }

        return nums[left];
    }

    public int findMax(int[] nums) {
        int left = 0;
        int right = nums.length - 1;

        while (left < right) {
            int mid = (right + left) >> 1;

            if (nums[mid] > nums[left]) {
                left = mid;
            } else if (nums[mid] < nums[left]) {
                right = mid - 1;
            } else {
                left++; // nums[mid] == nums[left], reduce the search space
            }
        }

        return nums[left];
    }


    public static void main(String[] args) {
        FindMinInRotatedSortedArray findMinInRotatedSortedArray = new FindMinInRotatedSortedArray();
        int[] nums = { 3, 4, 5, 6, 1, 2 };
        System.out.println(findMinInRotatedSortedArray.findMin(nums)); // 1
        System.out.println(findMinInRotatedSortedArray.findMax(nums)); // 6
        int []nums2 = {2, 2, 2, 3};
        System.out.println(findMinInRotatedSortedArray.findMin(nums2)); // 0
        System.out.println(findMinInRotatedSortedArray.findMax(nums2)); // 2
    }
}

File: DSA/Python/Questions/BinarySearch/FindMinInRotatedSortedArray.py
"""
# Question: Find Minimum in Rotated Sorted Array
# Link: https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/

# Problem Statement:
# Given a sorted array that has been rotated between 1 and n times, find the minimum element.
# The array does not contain duplicates.

# Example:
# Input: nums = [3,4,5,1,2]
# Output: 1
# Explanation: Original array was [1,2,3,4,5] and it was rotated 3 times.
"""

from typing import List


class FindMinInRotatedSortedArray:
    def find_min_binary(self, nums: List[int]) -> int:
        """
        Binary search approach
        Time Complexity: O(log n)
        Space Complexity: O(1)
        """
        left, right = 0, len(nums) - 1

        while left < right:
            mid = (left + right) >> 1

            if nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid

        return nums[left]

    def find_min_linear(self, nums: List[int]) -> int:
        """
        Linear search approach (for comparison)
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        for i in range(1, len(nums)):
            if nums[i] < nums[i - 1]:
                return nums[i]
        return nums[0]

    def find_min_divide_conquer(self, nums: List[int]) -> int:
        """
        Divide and conquer approach
        Time Complexity: O(log n)
        Space Complexity: O(log n) due to recursion
        """

        def find_min_recursive(left: int, right: int) -> int:
            if left == right:
                return nums[left]

            mid = (left + right) >> 1

            if nums[mid] > nums[right]:
                return find_min_recursive(mid + 1, right)
            return find_min_recursive(left, mid)

        return find_min_recursive(0, len(nums) - 1)


def main():
    test_cases = [
        {"nums": [3, 4, 5, 1, 2], "expected": 1},
        {"nums": [4, 5, 6, 7, 0, 1, 2], "expected": 0},
        {"nums": [11, 13, 15, 17], "expected": 11},
        {"nums": [2, 1], "expected": 1},
    ]

    solution = FindMinInRotatedSortedArray()
    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"Input: {test['nums']}")

        result_binary = solution.find_min_binary(test["nums"])
        result_linear = solution.find_min_linear(test["nums"])
        result_divide_conquer = solution.find_min_divide_conquer(test["nums"])

        print(f"Binary Search Result: {result_binary}")
        print(f"Linear Search Result: {result_linear}")
        print(f"Divide & Conquer Result: {result_divide_conquer}")
        print(f"Expected: {test['expected']}")

        assert (
            result_binary == result_linear == result_divide_conquer == test["expected"]
        )


if __name__ == "__main__":
    main()
